lib/charms/opensearch/v0/opensearch_relation_provider.py

```diff
@@ -0,0 +1,254 @@
+        network endpoints is unnecessary.
+        """
+        if not self.unit.is_leader():
+            return
+        if not self.opensearch.is_node_up():
+            event.defer()
+            return
+
+        try:
+            extra_user_roles = json.loads(event.extra_user_roles)
+        except (json.decoder.JSONDecodeError, TypeError) as err:
+            # TODO document what a client application would need to provide to make this work.
+            logger.error(err)
+            self.unit.status = BlockedStatus(ClientRelationBadRoleRequestMessage)
+            return
+
+        username = self._relation_username(event.relation)
+        hashed_pwd, pwd = generate_hashed_password()
+        try:
+            self.create_opensearch_users(username, hashed_pwd, extra_user_roles)
+        except OpenSearchUserMgmtError as err:
+            logger.error(err)
+            self.unit.status = BlockedStatus(str(err))
+            return
+
+        rel_id = event.relation.id
+        # Share the credentials and updated connection info with the client application.
+        self.database_provides.set_credentials(rel_id, username, pwd)
+        self.update_endpoints(event.relation)
+        self.database_provides.set_version(rel_id, self.opensearch.version)
+
+    def create_opensearch_users(
+        self, username: str, hashed_pwd: str, access_control: Dict[str, List[str]]
+    ):
+        """Creates necessary opensearch users and permissions for this relation.
+
+        Args:
+            username: Username to be created
+            hashed_pwd: the hash of the password to be assigned to the user
+            access_control: A dict of roles, permissions, and action groups to be applied to the
+                user. A new role will be created to contain the requested permissions and action
+                groups.
+        """
+        roles = access_control.get("roles")
+        permissions = access_control.get("permissions")
+        action_groups = access_control.get("action_groups")
+
+        if permissions or action_groups:
+            # combine action groups and perms into a new role for this relation.
+            try:
+                self.user_manager.create_role(
+                    role_name=username,
+                    permissions=permissions,
+                    action_groups=action_groups,
+                )
+                roles.append(username)
+            except OpenSearchUserMgmtError as err:
+                logger.error(err)
+                raise
+
+        try:
+            self.user_manager.create_user(
+                username,
+                roles,
+                hashed_pwd,
+            )
+            if roles:
+                self.user_manager.patch_user(
+                    username,
+                    [{"op": "replace", "path": "/opendistro_security_roles", "value": roles}],
+                )
+        except OpenSearchUserMgmtError as err:
+            logger.error(err)
+            raise
```

Comment:
Just to double check...In the case we error out at this point we don't want to revert the role creation in line 154, right?

Reply:
@deusebio good point! I think there is not immediate need for the cleanup, I think the idea here is to "cleanup later" (periodically on `update_status`) separate "cleanup routine" to simplify and centralize the flow